business/proxy_status.go

```diff
@@ -0,0 +1,129 @@
+package business
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/models"
+)
+
+type ProxyStatus struct {
+	k8s kubernetes.ClientInterface
+}
+
+func (in *ProxyStatus) GetWorkloadProxyStatus(workloadName, namespace string, workloadStatus models.WorkloadStatus) models.WorkloadStatus {
+	proxyStatuses, err := in.k8s.GetProxyStatus()
+	if err != nil {
+		return workloadStatus
+	}
+
+	ps := getProxyStatusByName(workloadName, namespace, proxyStatuses)
+	if ps != nil {
+		workloadStatus.ProxyStatus = castProxyStatus(*ps)
+	}
+	return workloadStatus
+}
+
+func (in *ProxyStatus) GetWorkloadsProxyStatuses(namespace string, workloadStatuses []models.WorkloadStatus) []models.WorkloadStatus {
```

Comment:
I'm trying to run your PR, but it seems there are some errors unnoticed due to this function and the one above (`GetWorkloadProxyStatus`), and in spite of this, the health shows as "proxy is synced" in the UI, which seems wrong.

In kiali logs I see: `Error while fetching app health: unknown (get pods istiod-5b9d44c58b-l6nzx)`. Not sure why I have this (I'll investigate more), but anyway, I think here and above you shouldn't ignore the error.

Perhaps instead of taking a `workloadStatus` as param and send it mutated as return, you should just return `[]ProxyStatus` (or map of apps+`[]ProxyStatus`) + the error, and let the caller (`health.go`) handle that error?

Reply:
> in spite of this, the health shows as "proxy is synced" in the UI, which seems wrong.
Good point!

The main problem here is that you need to allow new permissions to your cluster. Specifically, allow GET to `pod/proxy` subresource. I am now working on the operator side to make the change.

> Perhaps instead of taking a workloadStatus as param and send it mutated as return, you should just return []ProxyStatus (or map of apps+[]ProxyStatus) + the error, and let the caller (health.go) handle that error?

It sound pretty obvious as you said now. Since the logic in this file is only about ProxyStatus, why should we add an extra responsability to know about the WorkloadStatus. Yep, like it.

> PS: also in general, it would be nice to embrace using as much as possible pure functions whenever it's possible (and easy) :-)
Good to connect with the thoughts in the article. If you have some more, happy to exchange and read.

Reply:
> > PS: also in general, it would be nice to embrace using as much as possible pure functions whenever it's possible (and easy) :-)

> Good to connect with the thoughts in the article. If you have some more, happy to exchange and read.

For sure, open a parenthesis (and do some underrated value of code-reviews :-) ). That's one of the principle of functional programming ... We're using go, which isn't a functional-style language so we could just say f*** & YOLO but nonetheless I think the benefits of such principles are valid although we're not forced to applying them.

The article says:

-    They’re easier to reason about
-    They’re easier to combine
-    They’re easier to test
-    They’re easier to debug
-    They’re easier to parallelize

To me, in this particular case the main point is the first, easier to reason about:

Simply, just looking at this function signature `func GetWorkloadProxyStatus(workloadName, namespace string) []ProxyStatus` , as a caller, it looks very obvious: it will find proxy statuses out of my workload & namespace. No doubt.

Whereas `func GetWorkloadProxyStatus(workloadName, namespace string, workloadStatus WorkloadStatus) WorkloadStatus` works by side-effect. Just looking at the function signature isn't enough, as a caller, I'll be asking myself: what do I have to pass as workloadStatus input? Some empty object, something pre-initialized with important values that are going to be used internally? It's less clear, I'll have to step into it to make sure.

Also as you put it, not having to deal with responsibility of `WorkloadStatus` makes it simpler, and will still be working if we want to use the same function in different contexts.

Reply:
`business` package has responsability to populate the model.
So, I agree that perhaps it's clear to let to do
```
workloadStatus.ProxyStatus = in.businessLayer.ProxyStatus.GetWorkloadProxyStatus(workload, namespace)
```
Instead to populated it inside the method which is less clear and it tends to add complexity when we want to make changes.

I'm more pragmatic related to functional style: if it helps to simplify something, perfect.

But side effects are not bad in all scenarios and I tend to not generalize.